File: arduino/FlipskyFT85BD/FlipskyFT85BD.cpp
/**
 * FlipskyFT85BD.cpp
 * =================
 * Implementation of the FlipskyFT85BD Arduino library.
 * See FlipskyFT85BD.h for full documentation.
 *
 * License: MIT
 */

#include "FlipskyFT85BD.h"

// ── Protocol constants ──────────────────────────────────────────────────────

static constexpr uint8_t STX = 0xAA;
static constexpr uint8_t ETX = 0xDD;

static constexpr uint8_t CMD_OBTAIN_DATA = 0x00;
static constexpr uint8_t CMD_CAN_OVER    = 0x10;
static constexpr uint8_t CMD_KEEP_LIVE   = 0x19;
static constexpr uint8_t CMD_SET_SPEED   = 0x27;

// ── Error name table ────────────────────────────────────────────────────────

static const char* const ERROR_NAMES[] = {
    "NONE",
    "PHASE_A_OVER_CURRENT", "PHASE_B_OVER_CURRENT", "PHASE_C_OVER_CURRENT",
    "PHASE_A_SENSOR",       "PHASE_B_SENSOR",        "PHASE_C_SENSOR",
    "PHASE_SUM_NOT_ZERO",
    "BUS_UNDER_VOLTAGE",    "BUS_OVER_VOLTAGE",
    "MOSFET_OVERHEAT",      "MOTOR_OVERHEAT",        "MOSFET_TEMP_SENSOR",
    "WATCHDOG_RESET",       "FLASH_CORRUPT",          "MCU_UNDER_VOLTAGE",
    "MOTOR_TEMP_SENSOR",    "MOTOR_BLOCKED",          "DRIVER_FAULT",
    "PHASE_LOSS",           "BUS_OVER_CURRENT"
};
static constexpr uint8_t ERROR_NAMES_COUNT =
    sizeof(ERROR_NAMES) / sizeof(ERROR_NAMES[0]);

const char* FlipskyFT85BD::Telemetry::errorName() const {
    if (error_code < ERROR_NAMES_COUNT) return ERROR_NAMES[error_code];
    return "UNKNOWN";
}

// ── Constructor / begin ─────────────────────────────────────────────────────

FlipskyFT85BD::FlipskyFT85BD(HardwareSerial& serial,
                               uint32_t baud,
                               int8_t   rxPin,
                               int8_t   txPin,
                               uint16_t readTimeoutMs)
    : _serial(serial)
    , _baud(baud)
    , _rxPin(rxPin)
    , _txPin(txPin)
    , _readTimeoutMs(readTimeoutMs)
{}

void FlipskyFT85BD::begin() {
#if defined(ESP32)
    if (_rxPin >= 0 && _txPin >= 0) {
        _serial.begin(_baud, SERIAL_8N1, _rxPin, _txPin);
    } else {
        _serial.begin(_baud, SERIAL_8N1);
    }
#else
    // Arduino Mega / others: pin remapping not supported here
    _serial.begin(_baud);
#endif
}

// ── CRC-16 Modbus ───────────────────────────────────────────────────────────

uint16_t FlipskyFT85BD::_crc16(const uint8_t* data, uint8_t len) {
    uint16_t crc = 0xFFFF;
    for (uint8_t i = 0; i < len; i++) {
        crc ^= (uint16_t)data[i];
        for (uint8_t j = 0; j < 8; j++) {
            crc = (crc & 1) ? ((crc >> 1) ^ 0xA001) : (crc >> 1);
        }
    }
    return crc;
}

// ── Frame builder ───────────────────────────────────────────────────────────

uint8_t FlipskyFT85BD::_buildFrame(const uint8_t* info, uint8_t info_len) {
    // frame = STX | DLEN | INFO... | CRC_hi | CRC_lo | ETX
    uint16_t crc = _crc16(info, info_len);
    uint8_t  idx = 0;
    _buf[idx++] = STX;
    _buf[idx++] = info_len;
    for (uint8_t i = 0; i < info_len; i++) _buf[idx++] = info[i];
    _buf[idx++] = (uint8_t)(crc >> 8);
    _buf[idx++] = (uint8_t)(crc & 0xFF);
    _buf[idx++] = ETX;
    return idx;  // total frame length
}

// ── Low-level query ─────────────────────────────────────────────────────────

bool FlipskyFT85BD::_query(const uint8_t* req, uint8_t req_len, Telemetry& t) {
    // Flush stale data
    while (_serial.available()) _serial.read();

    _serial.write(req, req_len);

    // Accumulate bytes into rx_buf until we find a valid frame or time out
    uint8_t  rx_buf[64];
    uint8_t  rx_len  = 0;
    uint32_t deadline = millis() + _readTimeoutMs;

    while (millis() < deadline) {
        while (_serial.available() && rx_len < sizeof(rx_buf)) {
            rx_buf[rx_len++] = (uint8_t)_serial.read();
        }

        // Scan for STX
        for (uint8_t i = 0; i < rx_len; i++) {
            if (rx_buf[i] != STX) continue;

            // Need at least: STX DLEN ... CRC_hi CRC_lo ETX  = dlen + 5 bytes
            if (i + 1 >= rx_len) break;  // need DLEN byte
            uint8_t dlen = rx_buf[i + 1];
            uint8_t needed = dlen + 5;    // STX + DLEN + INFO + CRC_hi + CRC_lo + ETX

            if (i + needed > rx_len) break;  // incomplete frame

            const uint8_t* info = &rx_buf[i + 2];
            uint16_t crc_calc   = _crc16(info, dlen);
            uint16_t crc_rx     = ((uint16_t)rx_buf[i + 2 + dlen] << 8) |
                                              rx_buf[i + 3 + dlen];
            uint8_t etx         = rx_buf[i + 4 + dlen];

            if (etx != ETX || crc_calc != crc_rx) continue;

            return _parseTelemetry(info, dlen, t);
        }
        delay(1);
    }
    return false;  // timeout
}

// ── Telemetry parser ────────────────────────────────────────────────────────

bool FlipskyFT85BD::_parseTelemetry(const uint8_t* info, uint8_t info_len,
                                      Telemetry& t) {
    if (info_len < 1 || info[0] != CMD_OBTAIN_DATA) return false;

    const uint8_t* d = info + 1;
    uint8_t dlen     = info_len - 1;
    if (dlen < 28) return false;

    uint8_t i = 0;

    auto u8  = [&]() -> uint8_t  { return d[i++]; };
    auto i16 = [&]() -> int16_t  {
        int16_t v = (int16_t)((uint16_t)d[i] << 8 | d[i+1]);
        i += 2; return v;
    };
    auto i32 = [&]() -> int32_t  {
        int32_t v = (int32_t)((uint32_t)d[i]   << 24 | (uint32_t)d[i+1] << 16 |
                               (uint32_t)d[i+2] << 8  | (uint32_t)d[i+3]);
        i += 4; return v;
    };

    t.mcu_id          = u8();
    t.error_code      = u8();
    t.voltage_V       = i16() / 100.0f;
    t.batt_current_A  = i32() / 1000000.0f;
    t.motor_current_A = i32() / 1000000.0f;
    t.erpm            = i32();
    t.duty            = i16() / 10000.0f;
    t.temp_fet_C      = i16() / 100.0f;
    t.temp_motor_C    = i16() / 100.0f;
    t.cpu_load_pct    = i16() / 100.0f;
    // encoder_deg (4 bytes) intentionally skipped

    return true;
}

// ── Public API ──────────────────────────────────────────────────────────────

bool FlipskyFT85BD::getTelemetry(Telemetry& t, int16_t can_id) {
    uint8_t info[3];
    uint8_t info_len;

    if (can_id < 0) {
        info[0]  = CMD_OBTAIN_DATA;
        info_len = 1;
    } else {
        info[0]  = CMD_CAN_OVER;
        info[1]  = (uint8_t)can_id;
        info[2]  = CMD_OBTAIN_DATA;
        info_len = 3;
    }

    uint8_t frame_len = _buildFrame(info, info_len);
    return _query(_buf, frame_len, t);
}

void FlipskyFT85BD::setSpeed(int32_t erpm, int16_t can_id) {
    uint8_t info[7];
    uint8_t info_len;

    if (can_id < 0) {
        info[0]  = CMD_SET_SPEED;
        info[1]  = (uint8_t)(erpm >> 24);
        info[2]  = (uint8_t)(erpm >> 16);
        info[3]  = (uint8_t)(erpm >>  8);
        info[4]  = (uint8_t)(erpm);
        info_len = 5;
    } else {
        info[0]  = CMD_CAN_OVER;
        info[1]  = (uint8_t)can_id;
        info[2]  = CMD_SET_SPEED;
        info[3]  = (uint8_t)(erpm >> 24);
        info[4]  = (uint8_t)(erpm >> 16);
        info[5]  = (uint8_t)(erpm >>  8);
        info[6]  = (uint8_t)(erpm);
        info_len = 7;
    }

    uint8_t frame_len = _buildFrame(info, info_len);
    _serial.write(_buf, frame_len);
}

void FlipskyFT85BD::keepAlive() {
    uint8_t info[1] = { CMD_KEEP_LIVE };
    uint8_t frame_len = _buildFrame(info, 1);
    _serial.write(_buf, frame_len);
}

File: python/flipsky_ft85bd.py
"""
flipsky_ft85bd.py
=================
Python library for the Flipsky FT85BD dual ESC (and any Flipsky ESC using
the FTESC UART V1.4 protocol).

Protocol summary
----------------
Frame:  0xAA | DLEN | CMD | DATA... | CRC_hi | CRC_lo | 0xDD
CRC:    Modbus CRC-16  (poly 0xA001, init 0xFFFF)

Key commands
    0x00  OBTAIN_DATA_ONCE  – request 28-byte telemetry
    0x10  CAN_OVER          – forward cmd to slave via CAN bus
    0x19  KEEP_LIVE         – heartbeat (send every ≤500 ms while running)
    0x27  SET_SPEED         – int32 eRPM, big-endian

Hardware prerequisite
---------------------
In Flipsky ESC Tool → App Config → Input Signal Type must be set to **OFF**
on every channel you want to control via UART SET_SPEED.

Typical connection (FT85BD)
    Baud rate : 460800
    Master CAN ID : reported in telemetry mcu_id field (e.g. 251)
    Slave  CAN ID : reported in telemetry mcu_id field (e.g. 184)

Usage
-----
    import serial
    from flipsky_ft85bd import FlipskyFT85BD

    ser = serial.Serial('COM16', 460800, timeout=0.1)
    esc = FlipskyFT85BD(ser)

    t = esc.get_telemetry()          # master
    t = esc.get_telemetry(can_id=184) # slave via CAN

    esc.set_speed(5000)              # master forward
    esc.set_speed(-5000, can_id=184) # slave reverse
    esc.keep_alive()
"""

import struct
import time
from dataclasses import dataclass, field
from typing import Optional

try:
    import serial  # pyserial
except ImportError:
    serial = None  # allow import without pyserial for type-checking / docs

__all__ = ['FlipskyFT85BD', 'Telemetry', 'ERROR_NAMES']
__version__ = '1.0.0'

# ─── Protocol constants ────────────────────────────────────────────────────

STX = 0xAA
ETX = 0xDD

CMD_OBTAIN_DATA = 0x00
CMD_CAN_OVER    = 0x10
CMD_KEEP_LIVE   = 0x19
CMD_SET_SPEED   = 0x27

DEFAULT_BAUD    = 460800

ERROR_NAMES = {
    0:  'NONE',
    1:  'PHASE_A_OVER_CURRENT', 2:  'PHASE_B_OVER_CURRENT',
    3:  'PHASE_C_OVER_CURRENT', 4:  'PHASE_A_SENSOR',
    5:  'PHASE_B_SENSOR',       6:  'PHASE_C_SENSOR',
    7:  'PHASE_SUM_NOT_ZERO',   8:  'BUS_UNDER_VOLTAGE',
    9:  'BUS_OVER_VOLTAGE',     10: 'MOSFET_OVERHEAT',
    11: 'MOTOR_OVERHEAT',       12: 'MOSFET_TEMP_SENSOR',
    13: 'WATCHDOG_RESET',       14: 'FLASH_CORRUPT',
    15: 'MCU_UNDER_VOLTAGE',    16: 'MOTOR_TEMP_SENSOR',
    17: 'MOTOR_BLOCKED',        18: 'DRIVER_FAULT',
    19: 'PHASE_LOSS',           20: 'BUS_OVER_CURRENT',
}


# ─── Telemetry dataclass ───────────────────────────────────────────────────

@dataclass
class Telemetry:
    """Decoded telemetry from OBTAIN_DATA_ONCE (CMD 0x00)."""
    mcu_id:          int   = 0      # ESC's own CAN ID
    error_code:      int   = 0
    voltage_V:       float = 0.0
    batt_current_A:  float = 0.0
    motor_current_A: float = 0.0
    erpm:            int   = 0
    duty:            float = 0.0    # 0.0 – 1.0
    temp_fet_C:      float = 0.0
    temp_motor_C:    float = 0.0
    cpu_load_pct:    float = 0.0
    encoder_deg:     float = 0.0

    @property
    def error_name(self) -> str:
        return ERROR_NAMES.get(self.error_code, f'UNKNOWN({self.error_code})')

    @property
    def ok(self) -> bool:
        return self.error_code == 0

    def __str__(self) -> str:
        return (
            f'MCU={self.mcu_id}  err={self.error_name}  '
            f'V={self.voltage_V:.2f}V  '
            f'Ibatt={self.batt_current_A:.2f}A  Imot={self.motor_current_A:.2f}A  '
            f'eRPM={self.erpm:+d}  duty={self.duty*100:.1f}%  '
            f'Tfet={self.temp_fet_C:.1f}°C  Tmot={self.temp_motor_C:.1f}°C'
        )


# ─── Low-level frame helpers (usable without a class) ─────────────────────

def crc16(data: bytes) -> int:
    """Modbus CRC-16 (poly=0xA001, init=0xFFFF)."""
    crc = 0xFFFF
    for b in data:
        crc ^= b
        for _ in range(8):
            crc = (crc >> 1) ^ 0xA001 if crc & 1 else crc >> 1
    return crc


def build_frame(info: bytes) -> bytes:
    """Wrap an info payload in the FTESC UART frame."""
    c = crc16(info)
    return bytes([STX, len(info)]) + info + bytes([c >> 8, c & 0xFF, ETX])


def parse_frame(raw: bytes) -> Optional[bytes]:
    """
    Find and validate the first complete FTESC UART frame in *raw*.
    Returns the info payload bytes on success, or None.
    """
    idx = raw.find(STX)
    if idx == -1:
        return None
    try:
        dlen   = raw[idx + 1]
        info   = raw[idx + 2 : idx + 2 + dlen]
        crc_rx = (raw[idx + 2 + dlen] << 8) | raw[idx + 3 + dlen]
        etx    = raw[idx + 4 + dlen]
    except IndexError:
        return None
    if etx != ETX:
        return None
    if crc16(info) != crc_rx:
        return None
    return info


def parse_telemetry(info: bytes) -> Optional[Telemetry]:
    """Parse an info payload from CMD_OBTAIN_DATA response into a Telemetry."""
    if not info or info[0] != CMD_OBTAIN_DATA:
        return None
    d = info[1:]
    if len(d) < 28:
        return None

    i = 0
    def u8():
        nonlocal i; v = d[i]; i += 1; return v
    def i16():
        nonlocal i; v = struct.unpack_from('>h', d, i)[0]; i += 2; return v
    def i32():
        nonlocal i; v = struct.unpack_from('>i', d, i)[0]; i += 4; return v

    return Telemetry(
        mcu_id          = u8(),
        error_code      = u8(),
        voltage_V       = i16() / 100.0,
        batt_current_A  = i32() / 1_000_000.0,
        motor_current_A = i32() / 1_000_000.0,
        erpm            = i32(),
        duty            = i16() / 10_000.0,
        temp_fet_C      = i16() / 100.0,
        temp_motor_C    = i16() / 100.0,
        cpu_load_pct    = i16() / 100.0,
        encoder_deg     = i32() / 1_000_000.0,
    )


# ─── High-level class ──────────────────────────────────────────────────────

class FlipskyFT85BD:
    """
    Driver for the Flipsky FT85BD dual ESC (FTESC UART V1.4 protocol).

    Parameters
    ----------
    serial_port : serial.Serial
        An already-opened pyserial port (or any object with .write()/.read()
        and .in_waiting).
    read_timeout : float
        Seconds to wait for a telemetry reply (default 0.15 s).

    Example
    -------
    >>> import serial
    >>> from flipsky_ft85bd import FlipskyFT85BD
    >>> ser = serial.Serial('COM16', 460800, timeout=0.1)
    >>> esc = FlipskyFT85BD(ser)
    >>> t = esc.get_telemetry()
    >>> print(t)
    """

    def __init__(self, serial_port, read_timeout: float = 0.15):
        self._ser = serial_port
        self._read_timeout = read_timeout

    # ── Frame factories ──────────────────────────────────────────────────

    def _obtain_data_frame(self, can_id: int = -1) -> bytes:
        if can_id < 0:
            return build_frame(bytes([CMD_OBTAIN_DATA]))
        return build_frame(bytes([CMD_CAN_OVER, can_id, CMD_OBTAIN_DATA]))

    def _set_speed_frame(self, erpm: int, can_id: int = -1) -> bytes:
        payload = struct.pack('>i', erpm)
        if can_id < 0:
            return build_frame(bytes([CMD_SET_SPEED]) + payload)
        return build_frame(bytes([CMD_CAN_OVER, can_id, CMD_SET_SPEED]) + payload)

    def _keep_live_frame(self) -> bytes:
        return build_frame(bytes([CMD_KEEP_LIVE]))

    # ── Serial helpers ───────────────────────────────────────────────────

    def _query(self, request: bytes) -> Optional[Telemetry]:
        """Send request, wait for a valid telemetry reply."""
        self._ser.reset_input_buffer()
        self._ser.write(request)
        deadline = time.monotonic() + self._read_timeout
        buf = b''
        while time.monotonic() < deadline:
            chunk = self._ser.read(self._ser.in_waiting or 1)
            if chunk:
                buf += chunk
                info = parse_frame(buf)
                if info is not None:
                    return parse_telemetry(info)
        return None

    # ── Public API ───────────────────────────────────────────────────────

    def get_telemetry(self, can_id: int = -1) -> Optional[Telemetry]:
        """
        Request telemetry from master (can_id=-1) or a CAN slave.

        Returns a Telemetry dataclass, or None if no response within timeout.
        """
        return self._query(self._obtain_data_frame(can_id))

    def set_speed(self, erpm: int, can_id: int = -1) -> None:
        """
        Command a speed in electrical RPM.

        Positive = forward, negative = reverse.
        ESC Tool → Input Signal Type must be OFF for this to work.
        """
        self._ser.write(self._set_speed_frame(erpm, can_id))

    def stop(self, can_id: int = -1) -> None:
        """Immediately command zero speed."""
        self.set_speed(0, can_id)

    def keep_alive(self) -> None:
        """
        Send heartbeat.  Must be called at least every 500 ms while motors
        are running, otherwise the ESC will stop the motor.
        """
        self._ser.write(self._keep_live_frame())
